### PR TITLE
Use singular helper when creating checksums

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Extracted from https://github.com/dotnet/roslyn/pull/72494.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 1]
```diff
@@ -74,6 +74,9 @@ public static Checksum Create(Checksum checksum1, Checksum checksum2)
     public static Checksum Create(Checksum checksum1, Checksum checksum2, Checksum checksum3)
         => Create(stackalloc[] { checksum1, checksum2, checksum3 });
 
+    public static Checksum Create(Checksum checksum1, Checksum checksum2, Checksum checksum3, Checksum checksum4)
+        => Create(stackalloc[] { checksum1, checksum2, checksum3, checksum4 });
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:78]: needed in linked pr.

File: src/Workspaces/Core/Portable/Serialization/SerializerService.cs [hunk 2]
```diff
@@ -205,7 +205,7 @@ private IOptionsSerializationService GetOptionsSerializationService(string langu
         => _lazyLanguageSerializationService.GetOrAdd(languageName, n => _workspaceServices.GetLanguageServices(n).GetRequiredService<IOptionsSerializationService>());
 
     public Checksum CreateParseOptionsChecksum(ParseOptions value)
-        => Checksum.Create(value, this);
+        => Checksum.Create((value, @this: this), static (tuple, writer) => tuple.@this.SerializeParseOptions(tuple.value, writer));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SerializerService.cs:208]: no need for the specialized helper for ParseOptions all the way in teh chcksum layer.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 3]
```diff
@@ -82,71 +85,36 @@ public static Checksum Create(ReadOnlySpan<Checksum> hashes)
...
     public static Checksum Create(ArrayBuilder<Checksum> checksums)
-    {
-        using var stream = SerializableBytes.CreateWritableStream();
-
-        using (var writer = new ObjectWriter(stream, leaveOpen: true))
+        => Create(checksums, static (checksums, writer) =>
         {
             foreach (var checksum in checksums)
                 checksum.WriteTo(writer);
-        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:92]: all these helpers now defer to the helper that takes a callback and passed it an ObjectWriter to write data into (which itself is then checksumed).  This is teh same logic as before, just with teh common code of creating the objectwriter and checksum'ing it pushed to the helper.

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs [hunk 4]
```diff
@@ -82,71 +85,36 @@ public static Checksum Create(ReadOnlySpan<Checksum> hashes)
...
-        using (var objectWriter = new ObjectWriter(stream, leaveOpen: true))
-        {
-            serializer.Serialize(value!, objectWriter, context, CancellationToken.None);
-        }
-
-        stream.Position = 0;
-        return Create(stream);
-    }
-
-    public static Checksum Create(ParseOptions value, ISerializerService serializer)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs:140]: this helper went away entirely.  the one place that was calling it just inlined its behavior at the callsite.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**